File: src/lib/stores/geozone.ts
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Marc Hoffmann (b14ckyy)

// INAV geozone config (read). See docs/active/GEOZONES.md.
//
// Mirrors the Rust `GeozoneConfig` (commands/geozone.rs). `loaded` is the last snapshot read from the
// FC, rendered on the 2D/3D maps + listed in the Airspace Manager panel. Download is always-on at INAV
// connect (the store updates when the reads complete) and cleared on disconnect. Geozones are an INAV
// ≥8.0 feature; on older firmware `has_geozones` is false and `zones` is empty. Editing/writing is
// Phase 2 (a working/dirty copy + Save-to-FC) and not implemented yet.

import { writable, derived, get } from 'svelte/store';
import { invoke } from '@tauri-apps/api/core';
import { settings } from '$lib/stores/settings';
import { mission, launchPoint, editMode, toDeg, hasLocation } from '$lib/stores/mission';
import { homePosition } from '$lib/stores/home';
import { resolveMissionAltitudes } from '$lib/helpers/terrainProfile';
import {
  checkMissionGeozones, EMPTY_MISSION_RESULT, type PathPt, type GeozoneMissionResult,
} from '$lib/helpers/geozoneMissionCheck';

/** Geozone shape (matches the firmware enum). */
export const GEOZONE_SHAPE_CIRCULAR = 0;
export const GEOZONE_SHAPE_POLYGON = 1;
/** Geozone type (matches the firmware enum). */
export const GEOZONE_TYPE_EXCLUSIVE = 0;
export const GEOZONE_TYPE_INCLUSIVE = 1;
/** Fence action (matches the firmware enum). */
export const GEOZONE_ACTION_NONE = 0;
export const GEOZONE_ACTION_AVOID = 1;
export const GEOZONE_ACTION_POSHOLD = 2;
export const GEOZONE_ACTION_RTH = 3;

/** One geozone vertex (lat/lon in degrees × 1e7). */
export interface GeoZoneVertex {
  lat: number;
  lon: number;
}

/** One geozone. `zone_type` 0 = exclusive (NFZ), 1 = inclusive (FZ). `shape` 0 = circular, 1 = polygon.
 *  `fence_action` 0 = none, 1 = avoid, 2 = pos-hold, 3 = RTH. Altitudes in cm: `min_alt_cm` 0 = ground,
 *  `max_alt_cm` 0 = no upper limit (∞). Circular zones carry `radius_cm` + a single centre vertex;
 *  polygons carry all corners and `radius_cm` is null. */
export interface GeoZone {
  id: number;
  zone_type: number;
  shape: number;
  min_alt_cm: number;
  max_alt_cm: number;
  is_sealevel_ref: boolean;
  fence_action: number;
  radius_cm: number | null;
  vertices: GeoZoneVertex[];
}

export interface GeozoneConfig {
  zones: GeoZone[];
  /** True when the geozone feature is available (INAV ≥8.0) — drives the UI's visibility. */
  has_geozones: boolean;
}

/** FC config limits (INAV `MAX_GEOZONES_IN_CONFIG` / `MAX_VERTICES_IN_CONFIG`). */
export const MAX_GEOZONES = 63;
export const MAX_VERTICES_TOTAL = 126;
/** Earth circumference (m) — for the web-mercator tile-width sizing of a freshly-added zone. */
const EARTH_CIRCUMFERENCE_M = 40075016.686;

/** Ground width (m) of one 256px web-mercator tile at the given latitude + zoom. New zones are sized
 *  in tiles so they look sensible at the current map scale (a circle = 2 tiles radius → fits a 4×4 tile
 *  grid; a polygon trapezoid ≈ 2×3 tiles). */
function tileWidthM(lat: number, zoom: number): number {
  return (EARTH_CIRCUMFERENCE_M * Math.cos((lat * Math.PI) / 180)) / 2 ** zoom;
}

/** Last snapshot read from the FC (the dirty-compare baseline). Null until first read / when not
 *  connected to a (geozone-capable) INAV FC. */
export const geozoneConfig = writable<GeozoneConfig | null>(null);

/** Editable working copy — drives the map overlays + the panel list so edits reflect live (mirrors
 *  `safehomeWorking`). Reset to the loaded snapshot on load / revert. */
export const geozoneWorking = writable<GeozoneConfig | null>(null);

/** Map edit-lock: when true the 2D map markers are editable (drag / popups). Toggled from the panel;
 *  defaults to locked so zones aren't moved by accident. Cleared on disconnect. */
export const geozoneEditing = writable<boolean>(false);

/** True when the working copy differs from the loaded snapshot (enables "Save to FC"). */
export const geozoneDirty = derived(
  [geozoneConfig, geozoneWorking],
  ([$loaded, $working]) =>
    !!$loaded && !!$working && JSON.stringify($loaded) !== JSON.stringify($working),
);

/** Read the full geozone config from the FC (INAV ≥8.0). Always called on connect (download
 *  always-on). On failure / non-INAV, clears to null. Resets the working copy to the fresh snapshot. */
export async function loadGeozoneConfig(): Promise<void> {
  try {
    const cfg = await invoke<GeozoneConfig>('geozone_read_all');
    geozoneConfig.set(cfg);
    geozoneWorking.set(structuredClone(cfg));
  } catch (e) {
    console.warn('[geozone] loadGeozoneConfig failed', e);
    geozoneConfig.set(null);
    geozoneWorking.set(null);
  }
}

/** "Save to FC": send the working copy as one batch + EEPROM + reboot. Geozones only apply after a
 *  reboot (INAV recomputes the internal zone structures at boot), so the FC restarts and the link drops
 *  — we do NOT re-read here; the reconnect handshake reloads the saved config. */
export async function saveGeozoneConfig(): Promise<void> {
  const cfg = get(geozoneWorking);
  if (!cfg) return;
  await invoke('geozone_write_all', { config: cfg });
}

/** Discard pending edits — reset the working copy to the loaded snapshot. */
export function revertGeozoneWorking(): void {
  const loaded = get(geozoneConfig);
  geozoneWorking.set(loaded ? structuredClone(loaded) : null);
}

/** Clear everything (on disconnect). */
export function clearGeozones(): void {
  geozoneConfig.set(null);
  geozoneWorking.set(null);
  geozoneEditing.set(false);
}

// ── Working-copy mutations (panel + map editing) ──────────────────────────────────────────────────
// All operate on `geozoneWorking`; "Save to FC" later writes the whole copy. Lat/lon are deg×1e7.

/** Update the matching zone in the working copy via a transform. */
function updateZone(id: number, fn: (z: GeoZone) => GeoZone): void {
  geozoneWorking.update((c) =>
    c ? { ...c, zones: c.zones.map((z) => (z.id === id ? fn(z) : z)) } : c,
  );
}

/** Lowest unused zone id (0..62), or null when all slots are full. */
export function nextFreeGeozoneId(): number | null {
  const cfg = get(geozoneWorking);
  const used = new Set((cfg?.zones ?? []).map((z) => z.id));
  for (let i = 0; i < MAX_GEOZONES; i++) if (!used.has(i)) return i;
  return null;
}

/** Add a new zone (circle or polygon) at the persisted map centre, sized in tiles for the current
 *  map zoom (circle = 2 tiles radius; polygon = a ~2×3 tile trapezoid, so it reads clearly as a
 *  polygon). Sane defaults otherwise (exclusive, action None, ground → no upper limit). Returns the
 *  new zone's id, or null if the table is full. */
export function addGeozone(shape: number): number | null {
  const id = nextFreeGeozoneId();
  if (id == null) return null;
  const map = get(settings).map;
  const [lat, lon] = map.center;
  const tw = tileWidthM(lat, map.zoom);
  // Local metres → deg×1e7 around the centre.
  const dLatE7 = (m: number) => Math.round((m / 111320) * 1e7);
  const dLonE7 = (m: number) => Math.round((m / (111320 * Math.cos((lat * Math.PI) / 180))) * 1e7);
  const latE7 = Math.round(lat * 1e7), lonE7 = Math.round(lon * 1e7);

  let vertices: GeoZoneVertex[];
  let radius_cm: number | null = null;
  if (shape === GEOZONE_SHAPE_CIRCULAR) {
    vertices = [{ lat: latE7, lon: lonE7 }];
    radius_cm = Math.round(2 * tw * 100); // 2 tiles radius (cm)
  } else {
    // Trapezoid ≈ 2 tiles wide (bottom) × 3 tiles tall, narrower at the top. CCW order; the save step
    // re-normalises winding anyway.
    const halfH = 1.5 * tw, wBottom = 1.0 * tw, wTop = 0.6 * tw;
    vertices = [
      { lat: latE7 - dLatE7(halfH), lon: lonE7 - dLonE7(wBottom) }, // bottom-left
      { lat: latE7 - dLatE7(halfH), lon: lonE7 + dLonE7(wBottom) }, // bottom-right
      { lat: latE7 + dLatE7(halfH), lon: lonE7 + dLonE7(wTop) },    // top-right
      { lat: latE7 + dLatE7(halfH), lon: lonE7 - dLonE7(wTop) },    // top-left
    ];
  }
  const zone: GeoZone = {
    id,
    zone_type: GEOZONE_TYPE_EXCLUSIVE,
    shape,
    min_alt_cm: 0,
    max_alt_cm: 0,
    is_sealevel_ref: false,
    fence_action: GEOZONE_ACTION_NONE,
    radius_cm,
    vertices,
  };
  geozoneWorking.update((c) => (c ? { ...c, zones: [...c.zones, zone] } : c));
  return id;
}

/** Delete a zone from the working copy. */
export function deleteGeozone(id: number): void {
  geozoneWorking.update((c) => (c ? { ...c, zones: c.zones.filter((z) => z.id !== id) } : c));
}

export function setGeozoneType(id: number, zoneType: number): void {
  updateZone(id, (z) => ({ ...z, zone_type: zoneType }));
}
export function setGeozoneAction(id: number, action: number): void {
  updateZone(id, (z) => ({ ...z, fence_action: action }));
}
export function setGeozoneAlts(id: number, minCm: number, maxCm: number): void {
  updateZone(id, (z) => ({ ...z, min_alt_cm: minCm, max_alt_cm: maxCm }));
}
export function setGeozoneSealevel(id: number, sealevel: boolean): void {
  updateZone(id, (z) => ({ ...z, is_sealevel_ref: sealevel }));
}

/** Move one vertex (deg×1e7). For a circle this moves the centre (vertex 0). */
export function setGeozoneVertex(id: number, index: number, latE7: number, lonE7: number): void {
  updateZone(id, (z) => ({
    ...z,
    vertices: z.vertices.map((v, i) => (i === index ? { lat: latE7, lon: lonE7 } : v)),
  }));
}

/** Insert a vertex (polygon only) after `afterIndex` — used by the map's edge midpoint handles. */
export function insertGeozoneVertex(id: number, afterIndex: number, latE7: number, lonE7: number): void {
  updateZone(id, (z) => {
    if (z.shape === GEOZONE_SHAPE_CIRCULAR) return z;
    const vertices = z.vertices.slice();
    vertices.splice(afterIndex + 1, 0, { lat: latE7, lon: lonE7 });
    return { ...z, vertices };
  });
}

/** Remove a polygon vertex (kept ≥3). */
export function removeGeozoneVertex(id: number, index: number): void {
  updateZone(id, (z) => {
    if (z.shape === GEOZONE_SHAPE_CIRCULAR || z.vertices.length <= 3) return z;
    return { ...z, vertices: z.vertices.filter((_, i) => i !== index) };
  });
}

/** Set a circular zone's radius (cm). */
export function setGeozoneRadius(id: number, radiusCm: number): void {
  updateZone(id, (z) => ({ ...z, radius_cm: Math.max(0, Math.round(radiusCm)) }));
}

// ── Mission ↔ geozone safety check (hints only; see helpers/geozoneMissionCheck.ts) ───────────────
/** Result of checking the planned mission against the FC geozones — drives the mission-editor warning
 *  bar + the red 2D/3D violation overlays. Recomputed (debounced) whenever an input changes. */
export const geozoneMissionResult = writable<GeozoneMissionResult>(EMPTY_MISSION_RESULT);

let checkGen = 0;
let checkTimer: ReturnType<typeof setTimeout> | undefined;
let checkMaxTimer: ReturnType<typeof setTimeout> | undefined;
async function recomputeMissionCheck(): Promise<void> {
  const cfg = get(geozoneWorking);
  if (!get(editMode) || !cfg?.has_geozones || cfg.zones.length === 0) {
    geozoneMissionResult.set(EMPTY_MISSION_RESULT);
    return;
  }
  const launch = get(launchPoint);
  const wps = get(mission).waypoints;
  const located = wps.some((w) => hasLocation(w.action) && !(w.lat === 0 && w.lon === 0));
  if (!launch && !located) {
    geozoneMissionResult.set(EMPTY_MISSION_RESULT);
    return;
  }
  // Reuse the exact mission altitude resolution (terrain-sampled launch ground + per-WP MSL) so the 3D
  // overlay sits on the mission line. `launchGround` (terrain MSL at launch) is the relM=0 reference.
  const gen = ++checkGen;
  const { alts, launchGround } = await resolveMissionAltitudes(wps, launch);
  if (gen !== checkGen) return; // superseded while awaiting terrain
  const base = launchGround ?? 0;

  const path: PathPt[] = [];
  if (launch) path.push({ lat: launch.lat, lon: launch.lng, relM: 0, altMsl: base });
  for (let i = 0; i < wps.length; i++) {
    const w = wps[i];
    if (!hasLocation(w.action) || (w.lat === 0 && w.lon === 0)) continue;
    const a = alts.get(i);
    if (!a) continue;
    path.push({ lat: toDeg(w.lat), lon: toDeg(w.lon), relM: a.altMsl - base, altMsl: a.altMsl });
  }
  const hp = get(homePosition);
  const home = hp.set ? { lat: hp.lat, lon: hp.lon } : null;
  geozoneMissionResult.set(checkMissionGeozones(cfg.zones, path, home, launchGround));
}

function runCheckNow(): void {
  if (checkTimer) { clearTimeout(checkTimer); checkTimer = undefined; }
  if (checkMaxTimer) { clearTimeout(checkMaxTimer); checkMaxTimer = undefined; }
  void recomputeMissionCheck();
}
// Debounced (150 ms) with a 400 ms max-wait so frequent telemetry re-emits (home/mission) can't keep
// resetting the timer and starve the recompute.
function scheduleMissionCheck(): void {
  if (checkTimer) clearTimeout(checkTimer);
  if (!checkMaxTimer) checkMaxTimer = setTimeout(runCheckNow, 400);
  checkTimer = setTimeout(runCheckNow, 150);
}

// Recompute on any input change. Module-level — the result store lives for the app lifetime and is
// cheap when not in edit mode (early return). The edit-mode toggle runs the check IMMEDIATELY so the
// violations appear the instant you start editing (not on the next map re-render); the rest are debounced.
editMode.subscribe(() => runCheckNow());
geozoneWorking.subscribe(scheduleMissionCheck);
mission.subscribe(scheduleMissionCheck);
launchPoint.subscribe(scheduleMissionCheck);
homePosition.subscribe(scheduleMissionCheck);

File: src-tauri/src/commands/geozone.rs
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Marc Hoffmann (b14ckyy)

// INAV geozone config (read). See docs/active/GEOZONES.md.
//
// `geozone_read_all` reads every geozone slot the FC can hold (id 0..62) plus each used zone's
// vertices, for the map overlay + the Airspace Manager panel list. Geozones are an INAV ≥8.0 feature
// (gated by `FeatureSet.geozones`); on older firmware / non-INAV links we return an empty,
// `has_geozones=false` config. Writing/editing (batch SET + EEPROM) is Phase 2 and not implemented yet.

use serde::{Deserialize, Serialize};
use tauri::State;

use crate::msp::{
    MSP2_INAV_GEOZONE, MSP2_INAV_GEOZONE_VERTEX, MSP2_INAV_SET_GEOZONE, MSP2_INAV_SET_GEOZONE_VERTEX,
    MSP_EEPROM_WRITE, MSP_SET_REBOOT,
};
use crate::scheduler::SchedulerHandle;
use crate::state::{ActiveProtocol, AppState};

/// Geozone slots the FC config can hold (`MAX_GEOZONES_IN_CONFIG`; ids 0..62).
const MAX_GEOZONES: u8 = 63;

const GEOZONE_SHAPE_CIRCULAR: u8 = 0;

/// One geozone vertex (lat/lon in degrees × 1e7).
#[derive(Serialize, Deserialize, Clone)]
pub struct GeoZoneVertex {
    pub lat: i32,
    pub lon: i32,
}

/// One geozone. `zone_type` 0 = exclusive (NFZ), 1 = inclusive (FZ). `shape` 0 = circular, 1 = polygon.
/// `fence_action` 0 = none, 1 = avoid, 2 = pos-hold, 3 = RTH. Altitudes in cm: `min_alt_cm` 0 = ground,
/// `max_alt_cm` 0 = no upper limit. For a circular zone `radius_cm` is set and `vertices` holds the
/// single centre point; for a polygon `radius_cm` is None and `vertices` holds all corners.
#[derive(Serialize, Deserialize, Clone)]
pub struct GeoZone {
    pub id: u8,
    pub zone_type: u8,
    pub shape: u8,
    pub min_alt_cm: i32,
    pub max_alt_cm: i32,
    pub is_sealevel_ref: bool,
    pub fence_action: u8,
    pub radius_cm: Option<u32>,
    pub vertices: Vec<GeoZoneVertex>,
}

/// Full geozone snapshot for the frontend.
#[derive(Serialize, Deserialize, Clone, Default)]
pub struct GeozoneConfig {
    pub zones: Vec<GeoZone>,
    /// True when the geozone feature is available (INAV ≥8.0) — drives the UI's visibility.
    pub has_geozones: bool,
}

/// Resolve the MSP scheduler handle, erroring for non-MSP / disconnected links.
fn msp_handle(proto: &Option<ActiveProtocol>) -> Result<&SchedulerHandle, String> {
    match proto.as_ref() {
        Some(ActiveProtocol::Msp(h)) => Ok(h),
        Some(_) => Err("FC is not running MSP (INAV)".into()),
        None => Err("Not connected".into()),
    }
}

/// Read all geozones + their vertices. Returns an empty `has_geozones=false` config when the firmware
/// lacks the feature (<8.0), so callers can always invoke it on INAV connect.
#[tauri::command(async)]
pub fn geozone_read_all(state: State<'_, AppState>) -> Result<GeozoneConfig, String> {
    let proto = state.protocol.lock().map_err(|e| e.to_string())?;
    let handle = msp_handle(&proto)?;

    let has_geozones = {
        let info = state.fc_info.lock().map_err(|e| e.to_string())?;
        info.as_ref()
            .and_then(|fc| fc.features.as_ref())
            .map(|f| f.geozones)
            .unwrap_or(false)
    };
    if !has_geozones {
        return Ok(GeozoneConfig { zones: Vec::new(), has_geozones: false });
    }

    let mut zones = Vec::new();
    for id in 0..MAX_GEOZONES {
        // Header resp = [id, type, shape, minAlt(4), maxAlt(4), isSealevelRef, fenceAction, vertexCount] = 14 bytes.
        let r = handle.msp_request(MSP2_INAV_GEOZONE, &[id])?;
        if r.len() < 14 {
            continue;
        }
        let shape = r[2];
        let vertex_count = r[13];
        if vertex_count == 0 {
            continue; // unused slot
        }

        let mut vertices = Vec::with_capacity(vertex_count as usize);
        let mut radius_cm: Option<u32> = None;
        if shape == GEOZONE_SHAPE_CIRCULAR {
            // Circle: a single vertex (centre); the radius is appended (resp = 14 bytes).
            let v = handle.msp_request(MSP2_INAV_GEOZONE_VERTEX, &[id, 0])?;
            if v.len() >= 10 {
                vertices.push(GeoZoneVertex {
                    lat: i32::from_le_bytes([v[2], v[3], v[4], v[5]]),
                    lon: i32::from_le_bytes([v[6], v[7], v[8], v[9]]),
                });
            }
            if v.len() >= 14 {
                radius_cm = Some(u32::from_le_bytes([v[10], v[11], v[12], v[13]]));
            }
        } else {
            // Polygon: vertexCount corners, each resp = [zoneId, vertexId, lat(4), lon(4)] = 10 bytes.
            for vi in 0..vertex_count {
                let v = handle.msp_request(MSP2_INAV_GEOZONE_VERTEX, &[id, vi])?;
                if v.len() >= 10 {
                    vertices.push(GeoZoneVertex {
                        lat: i32::from_le_bytes([v[2], v[3], v[4], v[5]]),
                        lon: i32::from_le_bytes([v[6], v[7], v[8], v[9]]),
                    });
                }
            }
        }

        zones.push(GeoZone {
            id: r[0],
            zone_type: r[1],
            shape,
            min_alt_cm: i32::from_le_bytes([r[3], r[4], r[5], r[6]]),
            max_alt_cm: i32::from_le_bytes([r[7], r[8], r[9], r[10]]),
            is_sealevel_ref: r[11] != 0,
            fence_action: r[12],
            radius_cm,
            vertices,
        });
    }

    eprintln!("[GEOZONE] read {} active zone(s)", zones.len());
    Ok(GeozoneConfig { zones, has_geozones: true })
}

/// "Save to FC": write the whole geozone config as a batch, then a single EEPROM write to persist.
/// Every slot id 0..62 is written — active zones with their data, all other slots cleared
/// (`vertexCount = 0`) so removed zones don't linger. The zone header is written BEFORE its vertices
/// (the FC's vertex handler branches on the stored shape to read the circle radius); polygon vertices
/// go out in ascending order; a circle writes its single centre vertex with the radius appended.
#[tauri::command(async)]
pub fn geozone_write_all(config: GeozoneConfig, state: State<'_, AppState>) -> Result<(), String> {
    let proto = state.protocol.lock().map_err(|e| e.to_string())?;
    let handle = msp_handle(&proto)?;

    for id in 0..MAX_GEOZONES {
        let zone = config.zones.iter().find(|z| z.id == id);
        match zone {
            Some(z) => {
                let circular = z.shape == GEOZONE_SHAPE_CIRCULAR;
                let vertex_count: u8 = if circular { 1 } else { z.vertices.len() as u8 };
                // Header [id, type, shape, minAlt(4), maxAlt(4), isSealevelRef, fenceAction, vertexCount].
                let mut h = Vec::with_capacity(14);
                h.push(id);
                h.push(z.zone_type);
                h.push(z.shape);
                h.extend_from_slice(&z.min_alt_cm.to_le_bytes());
                h.extend_from_slice(&z.max_alt_cm.to_le_bytes());
                h.push(z.is_sealevel_ref as u8);
                h.push(z.fence_action);
                h.push(vertex_count);
                handle.msp_request(MSP2_INAV_SET_GEOZONE, &h)?;

                if circular {
                    // [id, 0, lat(4), lon(4), radius(4)] — FC stores the radius as the hidden vertex 1.
                    let c = z.vertices.first().ok_or("circular geozone has no centre vertex")?;
                    let mut p = Vec::with_capacity(14);
                    p.push(id);
                    p.push(0);
                    p.extend_from_slice(&c.lat.to_le_bytes());
                    p.extend_from_slice(&c.lon.to_le_bytes());
                    p.extend_from_slice(&z.radius_cm.unwrap_or(0).to_le_bytes());
                    handle.msp_request(MSP2_INAV_SET_GEOZONE_VERTEX, &p)?;
                } else {
                    for (vi, v) in z.vertices.iter().enumerate() {
                        let mut p = Vec::with_capacity(10);
                        p.push(id);
                        p.push(vi as u8);
                        p.extend_from_slice(&v.lat.to_le_bytes());
                        p.extend_from_slice(&v.lon.to_le_bytes());
                        handle.msp_request(MSP2_INAV_SET_GEOZONE_VERTEX, &p)?;
                    }
                }
            }
            None => {
                // Clear the slot: header with vertexCount 0 marks it unused.
                let mut h = Vec::with_capacity(14);
                h.push(id);
                h.extend_from_slice(&[0u8; 11]); // type, shape, minAlt(4), maxAlt(4), isSealevelRef
                h.push(0); // fenceAction
                h.push(0); // vertexCount
                handle.msp_request(MSP2_INAV_SET_GEOZONE, &h)?;
            }
        }
    }

    handle.msp_request(MSP_EEPROM_WRITE, &[])?;
    eprintln!("[GEOZONE] saved {} active zone(s) to FC (EEPROM written)", config.zones.len());

    // Geozones MUST be applied via a reboot: INAV recomputes the internal zone structures only at boot,
    // so the EEPROM write alone doesn't take effect. INAV ACKs the reboot before restarting; the link
    // then drops (the frontend reconnects + re-reads on handshake), so a missing/late reply is fine.
    let _ = handle.msp_request(MSP_SET_REBOOT, &[]);
    eprintln!("[GEOZONE] reboot requested to apply geozones");
    Ok(())
}
